### DISCUSSION TITLE
Combining multiple WRF history files with combine_files ('kernel died')

### DISCUSSION THREAD

**Discussion by tlsw231** [pydata/xarray#6586] [Q&A] [answered]
I have around ~3000 WRF history files which I want to combine. 

Pangeo has a useful [link](https://gallery.pangeo.io/repos/NCAR/notebook-gallery/notebooks/Run-Anywhere/WRF/wrf_ex.html) but the example there has one file with all the temporal snapshots, which is different from my case.

The instructions [here](https://wrf-python.readthedocs.io/en/latest/basic_usage.html) seem more relevant and when I try to follow the example there the following code works if I include only a few history files (e.g., `files[0:5]`) in the append step. The variable `pbl` is then read in correctly. But If I include all the files the append step fails with a 'kernel died' message. Is there a way to create the `wrflist` variable without `append`? In the examples at the link, the entries in `wrflist` are written out manually but that is not an option with ~3000 files. 
```
# files is just a list of strings with all the WRF history filenames
wrflist=[]
[ wrflist.append(Dataset(i)) for i in files[0:5] ]  # this step works but fails if I include all the filenames
pbl = wrf.combine_files(wrflist,'PBLH',wrf.ALL_TIMES,method='cat')
```
Any ideas on why the append step is failing?

**Comment by keewis** [comment]
Is there a reason why you use `list.append` in a `list` comprehension? I'd use this instead:
```python
wrflist = [Dataset(path) for path in files[0:5]]
```

not sure if that will fix your issue, though

**Reply by tlsw231** [reply]
I tried that and seem to be getting the same 'kernel died unexpectedly' message if I include all the filenames. I am trying this with 384 gb memory.

**Comment by dcherian** [comment]
Can you try [xwrf](https://xwrf.readthedocs.io/en/latest/) instead? You'll probably get better advice asking questions [there](https://github.com/ncar-xdev/xwrf/discussions)

**Reply by tlsw231** [reply]
Thanks, didn't even know about its existence! Will check it out.

**Reply by tlsw231** [reply]
Is it not available on conda? I am getting this error during installation after `conda install xwrf --channel conda-forge`
```
PackagesNotFoundError: The following packages are not available from current channels:
  - xwrf

Current channels:

  - https://conda.anaconda.org/conda-forge/linux-64
  - https://conda.anaconda.org/conda-forge/noarch
  - https://repo.anaconda.com/pkgs/main/linux-64
  - https://repo.anaconda.com/pkgs/main/noarch
  - https://repo.anaconda.com/pkgs/r/linux-64
  - https://repo.anaconda.com/pkgs/r/noarch
  ```

**Reply by andersy005** [reply]
@tlsw231, unfortunately, `xwrf` hasn't been published to conda-forge and PyPI (https://github.com/ncar-xdev/xwrf/issues/51) yet. Do you mind trying the main branch 

```bash
python -m pip install git+https://github.com/ncar-xdev/xwrf
```

**Reply by tlsw231** [reply]
Thanks, that worked.

**Reply by tlsw231** [reply]
How does one combine multiple flies in `xwrf`? I naively tried 
`ds=xr.open_mfdataset(files[0:5],parallel=True,concat_dim='Time',chunks='auto').xwrf.postprocess()`

and got this error:
`Could not find any dimension coordinates to use to order the datasets for concatenation`

[Edit: WRF defines its time dimensions weirdly (not datetime64 objects) which is probably why this is happening.]

**Reply by andersy005** [reply]
Not 💯 sure this works out of the box, but it's worth trying:


```python
import xarray as xr
dsets = [xr.open_dataset(file, chunks={}).xwrf.postprocess() for file in files]
ds = xr.concat(dsets, dim='Time')
```

**Reply by tlsw231** [reply]
I am having trouble getting it to work even with just one file, `xr.open_dataset(files[0],chunks={}).xwrf.postprocess()` gives me:

`arg must be a string, datetime, list, tuple, 1-d array, or Series`

**Reply by tlsw231** [reply]
If I use `.xwrf.postprocess(decode_times=False)`, the command goes through but then I lose the time information.

**Reply by tlsw231** [reply]
I confirmed the command below in `xwrf/postprocess.py` works as it should and returns a datetime64 object.
```
pd.to_datetime(xr.open_dataset(files[0]).Times.data.astype('str'),format='%Y-%m-%d_%H:%M:%S')
```

**Reply by andersy005** [reply]
this is more of an xwrf issue than an xarray one. Do you mind asking on the xwrf discussion forum: https://github.com/ncar-xdev/xwrf/discussions ?

**Reply by tlsw231** [reply]
Oh shoot, totally forgot about migrating the discussion there. Will do that.

**Comment by andersy005** [comment, marked as answer]
For anyone interested, the answer resides here: https://github.com/ncar-xdev/xwrf/discussions/78#discussioncomment-2720366
